File: c++/poo/empresaDesenvolvimento.cpp
#include <bits/stdc++.h>
using namespace std;

#define _ ios_base::sync_with_stdio(0);cin.tie(0)
#define endl '\n'
#define f first
#define s second

#define dbg(x) cout << #x << " = " << x << endl

typedef long long ll;

const int INF = 0x3f3f3f3f;
const ll LINF = 0x3f3f3f3f3f3f3f3fll;

class Project {
  public:
    int programmingRequirements;
    int designerRequirements;

};

class Employee {
  protected:
    int valuePerProject;
    int amountReceived;

  public:
    bool capacity(Project p) {
      return false;
    }

    void receiveReward() {
      this->amountReceived += valuePerProject;
    }

    int getAmoutReceived() {
      return this->amountReceived;
    }
};

class Developer : public Employee {
  int programmingSkills;

  public:
    Developer(int valuePerProject, int programmingSkills) {
      this->amountReceived = 0;
      this->valuePerProject = valuePerProject;
      this->programmingSkills = programmingSkills;
    }

    bool capacity(Project p) {
      return (this->programmingSkills >= p.programmingRequirements);
    }
};

class Designer : public Employee {
  int designSkills;

  public:
    Designer(int valuePerProject, int designSkills) {
      this->amountReceived = 0;
      this->valuePerProject = valuePerProject;
      this->designSkills = designSkills;
    }

    bool capacity(Project p) {
      return (this->designSkills >= p.designerRequirements);
    }
};

int main() { _;

  int valuePerProject, skill;

  cin >> valuePerProject >> skill;
  Developer developer = Developer(valuePerProject, skill);

  cin >> valuePerProject >> skill;
  Designer designer = Designer(valuePerProject, skill);

  int n; cin >> n;

  for(int i = 0; i < n; i++) {
    int p, d; cin >> p >> d;
    Project project;
    project.programmingRequirements = p;
    project.designerRequirements = d;

    if(developer.capacity(project) && designer.capacity(project)) {
      developer.receiveReward();
      designer.receiveReward();
    }

  }

  cout << "Programador: R$ " << developer.getAmoutReceived() << endl;
  cout << "Designer: R$ " << designer.getAmoutReceived() << endl;

  return 0;
}

File: python/poo/empresa_desenvolvimento.py
class Project:
  def __init__(self, requirement_programming: int, requirement_design: int) -> None:
    self.requirement_programming = requirement_programming
    self.requirement_design = requirement_design

class Employee:

  def __init__(self, value_per_project: int) -> None:
    self.__value_per_project = value_per_project
    self.__amount_received = 0

  @property
  def amount_received(self) -> int:
    return self.__amount_received
  
  def able(self, project: Project) -> bool:
    return False
  
  def receive_reward(self) -> None:
    self.__amount_received += self.__value_per_project

class Developer(Employee):
  def __init__(self, value_per_project: int, programming_ability: int)  -> None:
    super().__init__(value_per_project)
    self.programming_ability = programming_ability

  def able(self, project: Project) -> bool:
    return self.programming_ability >= project.requirement_programming
  
class Designer(Employee):
  def __init__(self, value_per_project: int, design_ability: int)  -> None:
    super().__init__(value_per_project)
    self.design_ability = design_ability

  def able(self, project: Project) -> bool:
    return self.design_ability >= project.requirement_design



def main() -> None:
  value_programming, skill_programming = map(int, input().split())
  value_design, skill_design = map(int, input().split())

  n = int(input())

  developer = Developer(value_programming, skill_programming)
  designer = Designer(value_design, skill_design)

  for _ in range(n):
    skill_p, skill_d = map(int, input().split())
    project = Project(skill_p, skill_d)

    if developer.able(project) and designer.able(project):
      developer.receive_reward()
      designer.receive_reward()
      

  print(f"Programador: R$ {developer.amount_received}")
  print(f"Designer: R$ {designer.amount_received}")


if __name__ == "__main__":
  main()